volatility3/framework/plugins/windows/pe_symbols.py

```diff
@@ -0,0 +1,437 @@
+        )
+
+        # special handling for the kernel
+        gather_kernel = os_module_name in wanted_modules
+
+        for index, mod in enumerate(
+            modules.Modules.list_modules(context, layer_name, symbol_table)
+        ):
+            try:
+                mod_name = mod.BaseDllName.get_string().lower()
+            except exceptions.InvalidAddressException:
+                continue
+
+            # to analyze, it must either be the kernel or a wanted module
+            if gather_kernel and index == 0:
+                mod_name = os_module_name
+            elif not mod_name.endswith(wanted_modules):
+                continue
+
+            # we won't find symbol information if we can't analyze the module
+            session_layer_name = modules.Modules.find_session_layer(
+                context, session_layers, mod.DllBase
+            )
+            if not session_layer_name:
+                continue
+
+            if mod_name not in found_modules:
+                found_modules[mod_name] = []
+
+            found_modules[mod_name].append(
+                (session_layer_name, mod.DllBase, mod.SizeOfImage)
+            )
+
+        return found_modules
+
+    @staticmethod
+    def get_wanted_process_modules(
+        context,
+        layer_name: str,
+        symbol_table: str,
+        wanted_symbols: Dict[str, List[str]],
+    ) -> Dict[str, Tuple[str, int, int]]:
+        """
+        Walks the process list and each process' VAD to determine the base address and size of wanted modules
+        """
+        proc_modules = {}
+
+        # create a tuple of module names for use with `endswith`
+        wanted_modules = tuple([key.lower() for key in wanted_symbols.keys()])
+
+        procs = pslist.PsList.list_processes(
+            context=context,
+            layer_name=layer_name,
+            symbol_table=symbol_table,
+        )
+
+        # walk the process list gathering processes that map the DLL(s) of interest
+        for proc in procs:
+            try:
+                proc_layer_name = proc.add_process_layer()
+            except exceptions.InvalidAddressException:
+                continue
+
+            for vad in proc.get_vad_root().traverse():
+                filepath = vad.get_file_name()
+                if not isinstance(filepath, str) or filepath.count("\\") == 0:
+                    continue
+
+                # track each module along with the process layer and range to find it
+                filename = filepath.lower().split("\\")[-1]
+                if filename.endswith(wanted_modules):
+                    if filename not in proc_modules:
+                        proc_modules[filename] = []
+
+                    proc_modules[filename].append(
+                        (proc_layer_name, vad.get_start(), vad.get_size())
+                    )
+
+        return proc_modules
+
+    @staticmethod
+    def resolve_wanted_process_symbols(
```

Comment:
Have you considered using partials instead of defining two functions that just sub in a particular input parameter?  You'd shift module collector to by the last parameter, then do:

```
from functools import partial
find_offsets = partial(PESymbols.find_function_offsets, context, config_path, layer_name, symbol_table, wanted_symbols)
```
and then you could call either `find_offsets(PESymbols.get_wanted_process_modules)` or `find_offsets(PESymbols.get_wanted_kernel_modules)`, and if you added methods later on, they'd also be able to fit into the same system?

Reply:
I never used partials before. I could explore this method after Black Hat as some plugins for the Black Hat <-> October timeframe will have a similar setup. It definitely seems cleaner

Reply:
Could you add a `FIXME` comment please, to ensure the things we want to do before October happen?